media-fonts/source-{pro,han-sans}: split source-pro, add new versions, bump EAPI, add zh-HK

tastytea opened the pull request:
The version bump fixes an incompatibility with =media-libs/freetype-2.12.1 ([bug 865103](https://bugs.gentoo.org/865103)). To be able to bump the fonts, they have to be split into their own packages:

> media-fonts/source-pro installs 3 font families. A while ago the serif
> and sans-serif fonts were renamed (and will be renamed again) while the
> monospace font kept the same name. Furthermore, the fonts now have the
> major version in their name.
> 
> The fonts will be split into their own packages and the old package
> will be masked since it isn't bump-able.

Upstream announcement: <https://github.com/adobe-fonts/source-serif/issues/77>

The new packages are slotted to provide backwards compatibility from now on. It will be impossible to install media-fonsts/source-code-pro and media-fonts/source-pro at the same time, because the files clash. That means it will be impossible to use the new Source Code Pro font together with the old Source Sans/Serif Pro fonts. I opted to not add media-fonts/source-{serif,sans}-pro fonts because i doubt it is worth the effort and the workaround (drop the .otf files into ~/.local/share/fonts) is easy enough.

I also bumped media-fonts/source-han-sans and added support for zh-HK.

tastytea commented:
dev-lang/nasm [requires Source Sans Pro and Source Code Pro](https://repo.or.cz/nasm.git/blob/68e3802b238b964900acac9422a70e295482243f:/doc/psfonts.ph). Looks like i have to add packages for the old fonts after all. Or maybe don't mask the old package and do a news item instead? Seems to be the simpler solution. :thinking: 

> USE_EXPAND group 'L10N' matches local USE flag: 'l10n_zh-HK'

Does that mean i should add it to `profiles/desc/l10n.desc` instead?

tastytea commented:
> dev-lang/nasm [requires Source Sans Pro and Source Code Pro](https://repo.or.cz/nasm.git/blob/68e3802b238b964900acac9422a70e295482243f:/doc/psfonts.ph). Looks like i have to add packages for the old fonts after all. Or maybe don't mask the old package and do a news item instead? Seems to be the simpler solution. thinking

I removed the mask for now and added a compatible version to media-fonts/source-sans. The plan is to stabilize that version quickly, fix the deps in dev-lang/nasm and mask media-fonts/source-pro afterwards.

> > USE_EXPAND group 'L10N' matches local USE flag: 'l10n_zh-HK'
> 
> Does that mean i should add it to `profiles/desc/l10n.desc` instead?

Yes, and that's what i did.

juippis reviewed (commented):
Why can't we just install all the families with "source-code-pro" single package? Put the hans variant behind "cjk" flag or something. Sounds much more simpler than this proposal. 

I'm just saying I feel like this will become abandonware and it's much simpler to do drive-by commits to source-code-pro if they're all contained in one package. Let me know what you think.

Also ping @thesamesam I saw you commented in the bug, if you have opinions.

tastytea commented:
> Why can't we just install all the families with "source-code-pro" single package? Put the hans variant behind "cjk" flag or something. Sounds much more simpler than this proposal.

Because Source Serif 3 and Source Sans 4 are different font families than Source Serif Pro and Source Sans Pro. The font famililies will change with every new major version. Source Code Pro and Source Han Sans will likely change font families too in the future.

dev-lang/nasm requires Source Sans Pro and Source Code Pro. It would need patching to work with Source Sans Pro 4 instead and would need patching again for Source Sans 5, 6 …

We *could* add aliases for the previous families to the current families, but that could have undesired consequences because the font metrics are different.

> I'm just saying I feel like this will become abandonware and it's much simpler to do drive-by commits to source-code-pro if they're all contained in one package. Let me know what you think.

I think it is simpler to to do drive-by changes if there aren't 3 or 4 different packages with different versions handled in one ebuild, and if there is no need to add new fontconfig aliases each time. I can add myself as proxied maintainer and take care of it?

tastytea commented:
- rebased
- dependency on `!>=media-libs/freetype-2.12.1` for source-pro
- added missing `<stabilize-allarches />`, added myself as proxied maintainer

juippis commented:
I gotta say I'm really not a fan of this current optfeature implementation, but I'll try to have enough time this weekend to dig in to the issue you linked, and figure some way of handling/merging this PR.

tastytea commented:
removed optfeature

juippis commented on media-fonts/source-pro/source-pro-20170111.ebuild:
```diff
@@ -28,6 +28,7 @@ RESTRICT="binchecks strip"
 RDEPEND="
 	media-libs/fontconfig
 	cjk? ( media-fonts/source-han-sans )
+	!>=media-libs/freetype-2.12.1
```
This needs a revbump. Could you also add it as the first element in RDEPEND? 

I imagine this kind of a blocker could be problematic, since it tells portage that if you install source-pro portage can uninstall the blocker (freetype in this case). Now most likely freetype will have rdeps which then prevents the removal, but I can imagine it'll lead to confusing portage error messages.

Instead I see two other options:

1: What if we didn't add this blocker, but instantly stabilized the new fonts you're adding here, and package.masked source-pro instead marking it to be removed? Then the package.mask message could say to switch using source-code-pro, source-sand and source-serif. 

2: Add the blocker to freetype and revbump freetype (I think you had this at some point?)

tastytea replied:
I prefer 1, because it looks like no one else ran into the bug so far.

juippis reviewed (commented):
Well you kind of misunderstood what I was after with the optfeature comment, I would've wanted to combine these packages like they were, just micromanaged it differently. But after some thought this way is acceptable and much saner when it comes to maintenance burden, than what I had in mind.

Could you also update fonts-meta package with this PR? I can too after this gets sorted. 
 - add source-code-pro to the first unconditional `|| ( )` block for a monospace font, then add source-han-sans to the `cjk? ( )` block. Doubt we need -sans or -serif to be added.

tastytea commented:
re-added optfeature. Should we replace source-pro with source-code-pro in fonts-meta after the stabilization? Since the meta package is stable on amd64 it seems to be the option with the least overhead.

juippis commented:
> re-added optfeature. Should we replace source-pro with source-code-pro in fonts-meta after the stabilization? Since the meta package is stable on amd64 it seems to be the option with the least overhead.

Yes please, just bump it to version 3 with ~ keywords. Let's stabilize it along with the added fonts here.

juippis reviewed (approved):
I'm gonna remove media-fonts/source-sans-2.020 as it looks out of place, it has unused but declared `cjk` IUSE and no optfeature. You can add it later if you wish after fixing these minor issues.

Okay, let's try it out!

tastytea commented:
> I'm gonna remove media-fonts/source-sans-2.020 as it looks out of place, it has unused but declared `cjk` IUSE and no optfeature. You can add it later if you wish after fixing these minor issues.

dev-lang/nasm depends on Source Sans Pro, 2.020 is the last version with that font family name. Fixed in <https://github.com/gentoo/gentoo/pull/27269>.
